### DISCUSSION TITLE
Some noob programmer asking about architectures

### DISCUSSION THREAD

**Discussion by IsaacMvmv** [ptitSeb/box64#1655] [General]
I have a rpi 4b 4gb with arm64 os installed. I want to use box86 and box64, BUT I need a different os for box86. Why?
If this project only translates from one architecture to another (Also more things, I know), why not creating a "mix" detecting x86 or x86_64 and executing that?
I don't know why, just asking
:D

**Comment by ptitSeb** [comment]
Because of "bitness", that is 32bits vs 64bits.
Box86 translate x86 code (so 32bits) and functions to native. To be able to call that translated functions, it needs to be at the same bitness. If not, than **all** structure, pointer and so on needs to be translated back and forth 32/64bits.
box86 and box64 doesn't do this conversion.
So
To use box64 you need a 64bits OS
To use box86 you need a 32bits OS

note that for 32bits on 64bits, some OS (like debian) can use Multiarch to have both architecture on the same OS.
other OS can use container to mix architecture.

(I may, at some point, create "box32" to run x86 software on 64bits OS, but for now, I'm not decided and the amount of work is huge, so not sure it'll exist one day)

**Comment by IsaacMvmv** [comment]
Tnks for explanation. Box32 will help a lot of users. But for now, i will adapt to all we have now

**Comment by hex007** [comment]
For box32 can we avoid multiarch requirements?

**Comment by ptitSeb** [comment]
That would be the point yes.

**Comment by alex-maiorov** [comment]
> Tnks for explanation. Box32 will help a lot of users. But for now, i will adapt to all we have now

I am not entirely sure, but you might be able to run box86 on a 64-bit OS provided you install 32-bit copies of every library in the system. If I am not mistaken then ARM can run 32 bit code on 64 bit processors in some backwards compatibility mode, like x86 does.

**Comment by ptitSeb** [comment]
> > Tnks for explanation. Box32 will help a lot of users. But for now, i will adapt to all we have now
> 
> I am not entirely sure, but you might be able to run box86 on a 64-bit OS provided you install 32-bit copies of every library in the system. If I am not mistaken then ARM can run 32 bit code on 64 bit processors in some backwards compatibility mode, like x86 does.

Yep, that's multiarch.
